### PR TITLE
[GOBBLIN-1485]Enable feature to get schema from writer schema when do hive registration

### PR DESCRIPTION
**Pull request by ZihanLi58**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1485

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
1. enable feature to get schema from writer schema instead of listing directory to get latest schema. 
2. When determine whether there is schema change, use schema version to compare to avoid special character issue. 

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Run test on parallel pipeline

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/HiveRegistrationUnitComparator.java [hunk 1]
```diff
@@ -142,12 +145,24 @@ public T compareIsStoredAsSubDirs() {
     return (T) this;
   }
 
+  private State extractSchemaVersion(State state) {
+    State newState = new State(state);
+    String schemaFromState = state.getProp(AvroSerdeUtils.AvroTableProperties.SCHEMA_LITERAL.getPropName());
+    if (schemaFromState != null && !schemaFromState.isEmpty()) {
+      String schemaVersion = AvroUtils.getSchemaCreationTime(new Schema.Parser().parse(schemaFromState));
+      if (schemaVersion != null && !schemaVersion.isEmpty()) {
```
Comment (sv2000) [inline gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/HiveRegistrationUnitComparator.java:153]: can we re-written using Strings.isNullOrEmpty().

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/HiveRegistrationUnitComparator.java [hunk 2]
```diff
@@ -142,12 +145,24 @@ public T compareIsStoredAsSubDirs() {
     return (T) this;
   }
 
+  private State extractSchemaVersion(State state) {
+    State newState = new State(state);
+    String schemaFromState = state.getProp(AvroSerdeUtils.AvroTableProperties.SCHEMA_LITERAL.getPropName());
+    if (schemaFromState != null && !schemaFromState.isEmpty()) {
```
Comment (sv2000) [inline gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/HiveRegistrationUnitComparator.java:151]: can be re-written using Strings.isNullOrEmpty()

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/HiveRegistrationUnitComparator.java [hunk 3]
```diff
@@ -142,12 +145,24 @@ public T compareIsStoredAsSubDirs() {
...
+      String schemaVersion = AvroUtils.getSchemaCreationTime(new Schema.Parser().parse(schemaFromState));
+      if (schemaVersion != null && !schemaVersion.isEmpty()) {
+         newState.removeProp(AvroSerdeUtils.AvroTableProperties.SCHEMA_LITERAL.getPropName());
```
Comment (sv2000) [inline gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/HiveRegistrationUnitComparator.java:154]: Why do we remove this property?

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java [hunk 4]
```diff
@@ -75,12 +78,17 @@
...
   public static final String INPUT_FORMAT_CLASS_KEY = "hiveOrcSerdeManager.inputFormatClass";
   public static final String DEFAULT_INPUT_FORMAT_CLASS = OrcInputFormat.class.getName();
+  public static final String WRITER_LATEST_SCHEMA = "writer.latest.schema";
 
   public static final String OUTPUT_FORMAT_CLASS_KEY = "hiveOrcSerdeManager.outputFormatClass";
   public static final String DEFAULT_OUTPUT_FORMAT_CLASS = OrcOutputFormat.class.getName();
 
   public static final String HIVE_SPEC_SCHEMA_READING_TIMER = "hiveOrcSerdeManager.schemaReadTimer";
 
+  public static final String HIVE_SPEC_SCHEMA_FROM_WRITER = "hiveOrcSerdeManager.getSchemaFromWriterSchema";
```
Comment (sv2000) [inline gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java:88]: Do we need to introduce a config? Can we always look for writer schema first and then default to file schema if it doesn't exist?

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java [hunk 5]
```diff
@@ -264,7 +272,18 @@ private void addSchemaProperties(Path path, HiveRegistrationUnit hiveUnit)
    *
    */
   protected void addSchemaPropertiesHelper(Path path, HiveRegistrationUnit hiveUnit) throws IOException {
-    TypeInfo schema = getSchemaFromLatestFile(path, this.fs);
+    TypeInfo schema;
+    if(props.getPropAsBoolean(HIVE_SPEC_SCHEMA_FROM_WRITER, DEFAULT_HIVE_SPEC_SCHEMA_FROM_WRITER)) {
+      try {
+        Preconditions.checkArgument(props.contains(WRITER_LATEST_SCHEMA));
+        Schema avroSchema = new Schema.Parser().parse(props.getProp(WRITER_LATEST_SCHEMA));
+        schema = TypeInfoUtils.getTypeInfoFromObjectInspector(new AvroObjectInspectorGenerator(avroSchema).getObjectInspector());
```
Comment (sv2000) [inline gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java:280]: There is a Avro->Orc schema converter utility method in AvroOrcSchemaConverter class. Can we use that, given that it is being used by GobblinOrcWriter to create the ORC file?

Comment (sv2000) [reply gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java:280]: The schema converter returns a TypeDescription object. Not sure if it is easy to go from TypeDescription to TypeInfo.

**Review by sv2000 — CHANGES_REQUESTED** [review]

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/HiveRegistrationUnitComparator.java (hunk 3 above)
Comment (ZihanLi58) [reply gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/HiveRegistrationUnitComparator.java:155]: The whole schema may contains special character, so when comparing whether we need to update the table, we remove this prop, but it's a new state object, so we are not removing this prop from the hive unit.

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java (hunk 4 above)
Comment (ZihanLi58) [reply gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java:88]: There are some use case that we have writer schema set but it's not the schema for this topic. i.e. in metadata pipeline, writer schema is the schema for GMCE. If we enable this feature for all, we'll hit issue in this case.

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java (hunk 5 above)
Comment (ZihanLi58) [reply gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java:280]: Yeah I was trying to do that. Several reasons here: 
1. AvroOrcSchemaConverter is now defined in gobblin-orc module, I don't think it make sense for us to introduce new dependency for hive registration module.
2.  It's doable to transfer TypeDescription to TypeInfo, but it's the same way that we need to use OrcUtils to create one objectInspector and get typeInfo there. As we are using the writer schema to get orcSchema, I think the two results should be the same? I verified one table and it looks good to me.
What do you think?

**Review by sv2000 — CHANGES_REQUESTED** [review]

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java (hunk 5 above)
Comment (sv2000) [reply gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/orc/HiveOrcSerDeManager.java:280]: Makes sense. Thanks for the explanation!

**Review by sv2000 — APPROVED** [review]
+1.

**Pull request merged**